File: dev_ws/src/learning_service_python/learning_service_python/person_info_server.py
"""
 @Description: 
 @version: v1.0
 @Author: HTY
 @Date: 2022-07-22 18:05:11
"""

from learning_interface.srv import PersonInfo

import rclpy
from rclpy.node import Node


class PersonInfoServer(Node):
    def __init__(self, node):
        super(PersonInfoServer, self).__init__(node)
        self.server = self.create_service(PersonInfo, "person_info_service", self.person_callback)
        self.get_logger().info("Service started")

    def person_callback(self, request, response):
        if request.name.upper() == "HTY":
            response.age = 21
            response.id = 3190105708
        elif request.name.lower() == "ZXA":
            response.age = 20
            response.id = 3190105598
        response.message = f"Get person info for {request.name}!  age: {response.age}  id: {response.id}"

        self.get_logger().info(response.message)
        return response


def main(args=None):
    rclpy.init(args=args)
    person_info_server = PersonInfoServer("person_info_server_node")
    rclpy.spin(person_info_server)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: dev_ws/src/learning_service_cpp/src/person_info_server.cpp
/*
 * @Description: ROS2 service服务器 接收自定义消息 cpp版本
 * @version: v1.0
 * @Author: HTY
 * @Date: 2022-07-22 20:37:15
 */

#include "rclcpp/rclcpp.hpp"
#include "learning_interface/srv/person_info.hpp"


class PersonInfoServer: public rclcpp::Node
{
private:
    rclcpp::Service<learning_interface::srv::PersonInfo>::SharedPtr server;

    void person_callback(const learning_interface::srv::PersonInfo::Request::SharedPtr & request,
                         const learning_interface::srv::PersonInfo::Response::SharedPtr & response)
    {
        if (request->name == "HTY")
        {
            response->age = 21;
            response->id = 3190105708;
        }
        else if (request->name == "ZXA")
        {
            response->age = 20;
            response->id = 3190105598;
        }
        response->message = "Get person info!";

        RCLCPP_INFO_STREAM(this->get_logger(), "name: " << request->name << " | age: " << (int) response->age << " | id: " << response->id);
    }

public:
    explicit PersonInfoServer(const std::string & node) : Node(node)
    {
        server = this->create_service<learning_interface::srv::PersonInfo>
                ("person_info_service",
                 std::bind(&PersonInfoServer::person_callback, this,
                           std::placeholders::_1, std::placeholders::_2));
        RCLCPP_INFO_STREAM(this->get_logger(), "Service started");
    }
};


int main(int argc, char const *argv[])
{
    rclcpp::init(argc, argv);
//    // 把节点的执行器变成多线程执行器, 避免死锁
//    rclcpp::executors::MultiThreadedExecutor executor;
//    executor.add_node(std::make_shared<PersonInfoServer>("person_info_server_node"));
//    executor.spin();
    rclcpp::spin(std::make_shared<PersonInfoServer>("person_info_server_node"));
    rclcpp::shutdown();
    return 0;
}